tests/nc-client/main.c

```diff
@@ -0,0 +1,151 @@
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include "events.h"
+#include "network.h"
+#include "sock.h"
+#include "warnp.h"
+
+struct senddata {
+	char * buffer;
+	int socket;
+	int conndone;
+};
+
+/* Forward declarations. */
+static int send_input(void *);
+static int callback_wrote(void * cookie, ssize_t lenwrit);
+static int callback_connected(void * cookie, int socket);
+
+/* Send data from stdin to a socket, or close the connection. */
+static int
+send_input(void * cookie)
+{
+	struct senddata * send = (struct senddata *)cookie;
+	size_t len = 0;
+	ssize_t read;
+
+	/* Ensure that the buffer is clear. */
```

Comment:
Since the buffer is guaranteed to be NULL when we first arrive here, why not put the buffer-freeing into callback_wrote?